### ISSUE TITLE
add support for EIP to single-node-asg module

### ISSUE THREAD

**Issue by ketzacoatl** [fpco/terraform-aws-foundation#205]
ATM the Single-Node ASG module includes support for an EBS volume that is associated with the ASG. Another use-case we run into is attaching an Elastic IP (EIP). Similar to the EBS volume, this is easy to attach to a single EC2 instance, but Terraform cannot do that for an EC2 instance that does not exist yet (eg when the EC2 instance is created by an ASG). We solve this in the same way as with the EBS volume - we give EC2 init enough info to attach the EIP when booting the instance.

## Requirements

* Update the `single-node-asg` module to add support for EIP in the same way the EBS volume is supported.
* Make the EIP optional, use a variable/local combination to enable/disable the resource with true/false --> 1/0
* Propose and review a method for attaching during init before implementing.
* Update an example project (in this repo) that uses the `single-node-asg` module, for TDD and to demonstrate how it works.

**Comment by Magicloud** [comment]
For this case, is aws_eip enough, or aws_eip_association is necessary? Quoting the note from Terraform doc.

> NOTE: aws_eip_association is useful in scenarios where EIPs are either pre-existing or distributed to customers or users and therefore cannot be changed.

**Comment by ketzacoatl** [comment]
That is a great question @Magicloud. I hope this helps to explain:

Terraform is used to create the EIP resource itself. With an ASG, Terraform creates the LC and ASG, but AWS creates the EC2 instances, so Terraform cannot associate/attach the EIP to an EC2 instance created by AWS ASG. That EIP created by Terraform is then floating and not attached to an EC2 instance, but in this design, we have a "single node ASG" - there is only one EC2 instance, and the EC2 instance can be provided init/shell/etc to attach specific resources. In the existing module design, we create an EBS volume to attach. The purpose of this issue is to add support for an EIP that is attached in the same way the EBS volume it attached.

**Comment by Magicloud** [comment]
Thanks. So the EIP is pre-existing (by Terraform).

**Comment by ketzacoatl** [comment]
Yes, you want to use `aws_eip` and not the `_association` resource. During EC2 init, use `aws` tools to associate the EIP with the EC2 instance.

**Comment by Magicloud** [comment]
I did something like attaching the EBS. But found that whether or not I want to associate the EIP, the instance has an Internet IP any way. Working on figuring out where is the problem.

**Comment by ketzacoatl** [comment]
@Magicloud, also make sure the ASG does not have a public IP, there is a module variable for that.

**Comment by Magicloud** [comment]
To confirm, you meant another requirement, right? Right now single-node-asg creates ASG with the option on to assign each instance a public IP.

**Comment by Magicloud** [comment]
Associating address via awscli gives me UnauthorizedOperation.

**Comment by Magicloud** [comment]
One more thing to change, I guess, subnets module, need to pass false to its public attribute.

And, have not figured this out, passing false to associate_public_ip_address of aws_launch_configuration does not set the launch conf in AWS to not associate, but default (Only assign a public IP address to instances launched in the default VPC and subnet).

Still digging.

**Comment by Magicloud** [comment]
After my testing, it is as expected, without public network accessing, aws cannot attach EBS or associate EIP.

So we cannot do "ASG does not have a public IP". At least a NAT for the whole group.
